enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,213 @@
+the OpenShift deployment model into the Edge where the power, memory and CPU budgets
+are reduced.
+
+While the default configuration makes sense to most of our customers in the Device
+Edge area, still our current CSI, CNI, and router components use 35% of the memory
+footprint and a significative CPU and power footprint. Making those components
+pluggable would allow customers who don't require advanced network features,
+persistent volumes, or ingress/routes to disable or switch those technologies
+and reduce the memory/cpu and disk/over-the-wire footprint.
+
+
+### User Stories
+
+An application developer needs the whole stack of components, creates an image
+that contains:
+1. microshift
+
+Nothing would change in this case, installing microshift, which becomnes a virtual
+package, installs the necessary dependencies:
+* microshift-core
+* microshift-networking-ovn
+* microshift-storage-topolvm
+* microshift-routes
+
+as well as any other dependencies.
+
+An application developer only needs advanced networking, but no persistent volumes,
+and no routes/ingress.  Creates an image that contains:
+1. microshift-core
+2. microshift-networking-ovn
+
+
+An application developer wants the bare minimum set of components creates an image
+with:
+1. microshift-core
+
+An application developer wanting bare minimum networking but persistent volumes
+creates an image with:
+1. microshift-core
+2. microshift-storage-topolvm
+
+### Goals
+* Allow networking, storage and router core components to be optionally
+  installed.
+
+* As we refactor and re-use the kustomize package used for loading user workloads,
+  add support for loading yamls from multiple directories, i.e.:
+  * `/usr/lib/microshift/manifests.d/user-app-1/`
+  * `/usr/lib/microshift/manifests.d/user-app-2/`
+  * `/etc/microshift/manifests.d/user-app-3/`
+  * `...`
+
+* Split the RPM packages to resemble those optional components.
+
+* Preserve previous behavior in terms of RPM package installation.
+
+### Non-Goals
+
+* Making DNS component optional.
+* Making the certificate management component optional.
+
+## Proposal
+
+To allow modularity of the described core components, the following changes
+would be introduced:
+
+Update the service manager to allow core components as kustomization yamls,
+in the same way we already support for user application manifests. i.e.:
+
+`/usr/lib/microshift/components.d/networking-ovn/kustomize.yaml`
+`/usr/lib/microshift/components.d/storage-topolvm/kustomize.yaml`
+`/usr/lib/microshift/components.d/openshift-router/kustomize.yaml`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
+  directory and extract metadata from the yamls to analyze dependencies between
+  core components at boot. Then start the right componments in the right order.
+
+* The kustomize component loader would be refactored so it can be used for this
+  purpose. We should be able to pre-process kustomize yamls to insert template
```

Comment:
Could we list those template vars in the enhancement? I might be mistaken, but I have a feeling that this templating will be a little different from what we have right now: right now we're substituting stuff like image, but I'm assuming that pluggable component image ref will be stored in the yaml itself (within the RPM - we'll need to make sure the RPMs are the same version so no one installs mismatched core and plugin).

I mean that if we consider ovnk, topolvm, router optional then we won't have much about them in the binary, right?

Reply:
True, we won't have to template in images anymore, so in some cases we may get by without config from MicroShift to plugin, e.g. with CSI. In the CNI case there'll be vars the plugin will need to render (e.g. MTU) and those that will need to come from MicroShift (like ServiceCIDR).

Reply:
We shall limit/standardize what config (e.g. serviceCIDR, networkCIDR) need to be provided from MicroShift to plugins, and what others shall be provided by plugin specific config file (e.g. MTU for ovnk). The ability to render manifests with plugin specific config is also desired (either pass it back to microshift which is responsible for rendering manifests, or do it in plugin specific binary).